File: 0502-ipo/0502-ipo.java
class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {

        int n=profits.length;     // Kisi ka bhi Size lelo dono Same hai
        int[][] proj=new int[n][2];     // Pair lelo

        for(int i=0;i<n;i++){
            proj[i][0]=capital[i];
            proj[i][1]=profits[i];       // Iss Structure mein daal ke karlo Pair
        }

        Arrays.sort(proj,(a,b)->a[0]-b[0]);        // Ascending mein Sort hojaega

        PriorityQueue<Integer> pq=new PriorityQueue<>(Collections.reverseOrder());     // Max Heap

        int idx=0;                             // Pair wley Array ke liye Index Ptr

        while(k-- > 0){

            while(idx<n){

                if(proj[idx][0]>w){
                    // Nahi aaskata hai
                    break;
                }

                // Agar aaskata hai tab
                pq.offer(proj[idx][1]);        // Profits k hisab se Store karlo

                idx++;                         // Inc kartey chlo
            }

            // Edge Case
            if(pq.isEmpty()){
                return w;
            }

            w=w+pq.poll();          // Max Profit wley ko lekey Add kardo w mein
        }
        return w;             // Final Max Capital
    }
}

File: 0502-ipo/0502-ipo.py
import heapq

class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:

        n=len(profits)             # Kisi ka bhi Size lelo dono Same hai
        proj=[]                    # Pair lelo

        for i in range(n):
            proj.append((capital[i],profits[i]))       # Iss Structure mein daal ke karlo Pair

        proj.sort()                # Ascending mein Sort hojaega

        pq=[]                      # Max Heap

        idx=0                      # Pair wley Array ke liye Index Ptr

        while k>0:

            while idx<n:

                if proj[idx][0]>w:
                    # Nahi aaskata hai
                    break

                # Agar aaskata hai tab
                heapq.heappush(pq,-proj[idx][1])       # Profits k hisab se Store karlo

                idx+=1                                  # Inc kartey chlo

            # Edge Case
            if len(pq)==0:
                return w

            w=w+(-heapq.heappop(pq))        # Max Profit wley ko lekey Add kardo w mein

            k-=1

        return w                    # Final Max Capital